Proposals/20200511-HVPA-design.md

```diff
@@ -0,0 +1,262 @@
+
+#### HPA domination
+HPA is a more macroscopic autoscaler than VPA, because HPA recommeds change in replicas while VPA recommends change in resources. Because of this, total CPU/memory resources recommended by HPA might tend to dominate under increasing load.
+
+*Mitigation* - Either do not use HPA by setting `.spec.hpa.deploy` to `false` , or set the threshold percentages for HPA accordingly, eg - above 100%. Another alternative is to rely on VPA tuning as explained [here](####vpa-tuning).
+
+## Design Details
+
+A user can provide an array of buckets described above in the [proposal](##proposal).
+For example, typical `scaleIntervals` would look like:
+```yaml
+spec:
+  scaleIntervals:
+  - maxCpu: "1",
+    maxMemory: "2G",
+    maxReplicas: "1",
+  - maxCpu: "20",
+    maxMemory: "30G",
+    maxReplicas: "5",
+```
+
+### Minimum required fields
+All the fields in a bucket - `maxCpu`, `maxMemory` and `maxReplicas` - are required to define a valid bucket. These fields are required to determine upper bound of the aggregate resources for that bucket.
+This upper bound of a bucket is considered as lower bound for the next bucket unless it is the first bucket in which case the lower bound is calculated using VPA's `minAllowed`s or `0`, and HPA's `minReplicas` or `1`.
+
+For subsequent buckets:
+* the minimum replicas is equal to one more than `maxReplicas` of previous bucket
+* minimum resource values are calculated such that aggregated maximum resources for previous bucket equals aggregated minimum resources for the bucket
+
+For the second bucket in the spec shown in above example, the minimum replicas is `2`, the minimum cpu is `500m` (`1 replica * 1 core / 2 replicas`) and the minimum memory is `1G` (`1 replica * 2G memory / 2 replicas`).
+
+### kube-apiserver scaling example:
+For kube-apiserver, a typical spec could look like:
+
+```yaml
+spec:
+  scaleIntervals:
+  - maxCpu: "500m",
+    maxMemory: "2G",
+    maxReplicas: "1",
+  - maxCpu: "1",
+    maxMemory: "4G",
+    maxReplicas: "2",
+  - maxCpu: "2",
+    maxMemory: "8G",
+    maxReplicas: "3",
+  - maxCpu: "4",
+    maxMemory: "16G",
+    maxReplicas: "4",
+  - maxCpu: "8",
+    maxMemory: "32G",
+    maxReplicas: "5",
+  - maxCpu: "16",
+    maxMemory: "64G",
+    maxReplicas: "6",
+```
+* Here HVPA scales kube-apiserver from 1 to 2 replicas only when aggregate VPA recommendations cross 500m for CPU or 2GB for memory
+* Similarly kube-apiserver is scaled out from 2 replicas if aggregate VPA recommendations cross maximum aggregate resources for second bucket, i.e., 2 CPU or 8GB memory
+* Scale down from 3 replicas happen only when aggregate VPA recommendations go below minimum aggregate resources for third bucket, i.e., 2 CPU or 8GB memory. These values are obtained using maximum aggreagte resource values from second bucket
```

Comment:
This forces it to flip back and forth with no hysteresis:
> scale out above 2 CPU / 8 GB
> scale in below 2 CPU / 8 GB

Reply:
We already have a `minChange` field in:
```
// ScaleType defines parameters for scaling
type ScaleType struct {
	UpdatePolicy UpdatePolicy `json:"updatePolicy,omitempty"`

	// MinChange is the minimum change in the resource on which HVPA acts
	// HVPA uses minimum of the Value and Percentage value
	MinChange ScaleParams `json:"minChange,omitempty"`

	// StabilizationDuration defines the minimum delay in minutes between 2 consecutive scale operations
	// Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h"
	StabilizationDuration *string `json:"stabilizationDuration,omitempty"`
}
```
This field can be set separately for scale up and down.
For eg. if it set to 0.2 CPU / 300M for both, then scaling wouldn't happen until aggregate recommendations are out of the range of 1.8 CPU / 7.7GB and 2.2 CPU / 8.3GB.
This provides stability when recommendations fluctuate around a value.
For a finer control, we can reintroduce `minValue` field in the bucket which was removed in [this](https://github.com/gardener/hvpa-controller/pull/71/commits/f7bcb341c6330ad0583a378dff390e753a6065ca#diff-2b0b880941da62632ec7726b98fe0754R139) commit after [this](https://github.com/gardener/hvpa-controller/pull/71#discussion_r427094513) discussion.
wdyt?
/cc @amshuman-kr